### PR TITLE
Security enhancement butriy edition

### PR DESCRIPTION
**Pull request by NikPaukov**

### CODE REVIEW

File: src/test/java/com/qunite/api/web/AuthenticationIntegrationTest.java [hunk 1]
```diff
@@ -150,4 +155,69 @@ private static Stream<Arguments> signInShouldCheckCredentials() {
...
+  @Sql("/users-create.sql")
+  void updatedUsername() throws Exception {
+    var token = getAccessToken("First", "asd");
+
+    var userUpdateDto = new UserUpdateDto();
+    userUpdateDto.setUsername("NEW USERNAME");
+    var json = new ObjectMapper().writeValueAsString(userUpdateDto);
+
+    mockMvc.perform(patch("/users/self").contentType(MediaType.APPLICATION_JSON).content(json)
+            .header("authorization", token))
```
Comment (vetterocki) [inline src/test/java/com/qunite/api/web/AuthenticationIntegrationTest.java:183]: you can insert HttpHeaders.AUTHORIZATION static import

**Review by vetterocki — CHANGES_REQUESTED** [review]

File: src/main/java/com/qunite/api/web/dto/auth/AuthenticationResponse.java [hunk 2]
```diff
@@ -2,9 +2,11 @@
 
 import lombok.AllArgsConstructor;
 import lombok.Data;
+import lombok.NoArgsConstructor;
 
 @Data
 @AllArgsConstructor
+@NoArgsConstructor
```
Comment (xd6d) [inline src/main/java/com/qunite/api/web/dto/auth/AuthenticationResponse.java:9]: For what reason?

File: src/test/java/com/qunite/api/web/AuthenticationIntegrationTest.java [hunk 3]
```diff
@@ -150,4 +155,69 @@ private static Stream<Arguments> signInShouldCheckCredentials() {
         Arguments.of("John@user.com", "asd"),
         Arguments.of("First", "dsa"));
   }
+
+  @Test
+  @DisplayName("Token shouldn't work when user deleted")
```
Comment (xd6d) [inline src/test/java/com/qunite/api/web/AuthenticationIntegrationTest.java:160]: fix English
when user have been deleted

File: src/test/java/com/qunite/api/web/AuthenticationIntegrationTest.java [hunk 4]
```diff
@@ -150,4 +155,69 @@ private static Stream<Arguments> signInShouldCheckCredentials() {
...
+
+    mockMvc.perform(delete("/users/self").header("authorization", token))
+        .andExpect(status().isNoContent());
+
+    mockMvc.perform(get("/users/self").header("authorization", token))
+        .andExpect(status().isForbidden());
+  }
+
+  @Test
+  @DisplayName("Token shouldn't work when username changed")
```
Comment (xd6d) [inline src/test/java/com/qunite/api/web/AuthenticationIntegrationTest.java:173]: fix English

File: src/test/java/com/qunite/api/web/AuthenticationIntegrationTest.java [hunk 5]
```diff
@@ -150,4 +155,69 @@ private static Stream<Arguments> signInShouldCheckCredentials() {
...
+    mockMvc.perform(patch("/users/self").contentType(MediaType.APPLICATION_JSON).content(json)
+            .header("authorization", token))
+        .andExpect(status().isOk());
+
+    mockMvc.perform(get("/users/self").header("authorization", token))
+        .andExpect(status().isForbidden());
+  }
+
+  @Test
+  @DisplayName("Token should work when username not changed")
```
Comment (xd6d) [inline src/test/java/com/qunite/api/web/AuthenticationIntegrationTest.java:191]: lol

File: lombok.config [hunk 6]
```diff
@@ -0,0 +1 @@
+lombok.copyableAnnotations += org.springframework.context.annotation.Lazy
```
Comment (vetterocki) [inline lombok.config:1]: delete this

File: src/main/java/com/qunite/api/security/JwtService.java [hunk 7]
```diff
@@ -4,31 +4,48 @@
...
 import java.util.Optional;
+import lombok.RequiredArgsConstructor;
 import org.springframework.beans.factory.annotation.Value;
+import org.springframework.context.annotation.Lazy;
 import org.springframework.stereotype.Component;
 
 @Component
+@RequiredArgsConstructor
 public class JwtService {
   @Value("${jwt.secret}")
```
Comment (vetterocki) [inline src/main/java/com/qunite/api/security/JwtService.java:18]: use @RequiredArgsConstructor(onConstructor = @__(@Lazy))

**Review by vetterocki — CHANGES_REQUESTED** [review]

File: src/test/java/com/qunite/api/web/AuthenticationIntegrationTest.java
Comment (xd6d) [inline src/test/java/com/qunite/api/web/AuthenticationIntegrationTest.java:1]: I have counted 3 different invocations of `new ObjectMapper()`.
Maybe introduce local variable?

**Review by xd6d — CHANGES_REQUESTED** [review]

File: src/main/java/com/qunite/api/security/JwtService.java (hunk 7 above)
Comment (xd6d) [reply src/main/java/com/qunite/api/security/JwtService.java:18]: According to Lombok API docs, the recommended way is onConstructor_ ={@Lazy} for JDK8+. For JDK7 is onConstructor = @__(@Lazy)

source: https://stackoverflow.com/a/70519537/20311078

File: lombok.config (hunk 6 above)
Comment (NikPaukov) [reply lombok.config:1]: No.

File: src/main/java/com/qunite/api/security/JwtService.java (hunk 7 above)
Comment (NikPaukov) [reply src/main/java/com/qunite/api/security/JwtService.java:18]: I don`t like annotating all dependencies with lazy, it`s better to move lombok config to src and fix the problem.

File: src/main/java/com/qunite/api/web/dto/auth/AuthenticationResponse.java (hunk 2 above)
Comment (NikPaukov) [reply src/main/java/com/qunite/api/web/dto/auth/AuthenticationResponse.java:9]: To readAccesToken in AccAuthenticationIntegrationTest.getAccessToken(). Besides that, I see no problem in dto having emptyConstructor.

File: src/test/java/com/qunite/api/web/AuthenticationIntegrationTest.java (hunk 5 above)
Comment (NikPaukov) [reply src/test/java/com/qunite/api/web/AuthenticationIntegrationTest.java:191]: Can you add more details to your claim?

File: src/main/resources/application.yml
Comment (xd6d) [inline src/main/resources/application.yml:1]: delete

File: src/test/java/com/qunite/api/web/AuthenticationIntegrationTest.java [hunk 8]
```diff
@@ -157,62 +160,61 @@ private static Stream<Arguments> signInShouldCheckCredentials() {
...
         .andExpect(status().isOk());
 
-    mockMvc.perform(get("/users/self").header("authorization", token))
+    mockMvc.perform(get("/users/self").header(HttpHeaders.AUTHORIZATION, token))
         .andExpect(status().isForbidden());
   }
 
   @Test
-  @DisplayName("Token should work when username not changed")
+  @DisplayName("Token should work when username has`nt been changed")
```
Comment (xd6d) [inline src/test/java/com/qunite/api/web/AuthenticationIntegrationTest.java:194]: fix typo

**Review by xd6d — CHANGES_REQUESTED** [review]

File: src/lombok.config
Comment (xd6d) [inline src/lombok.config:1]: mb move this to resources?

File: src/main/java/com/qunite/api/web/dto/auth/AuthenticationResponse.java (hunk 2 above)
Comment (xd6d) [reply src/main/java/com/qunite/api/web/dto/auth/AuthenticationResponse.java:9]: I do not think that adding code only and only for your tests is a good practise

File: src/main/java/com/qunite/api/security/JwtService.java (hunk 7 above)
Comment (vetterocki) [reply src/main/java/com/qunite/api/security/JwtService.java:18]: > I don`t like annotating all dependencies with lazy, it`s better to move lombok config to src and fix the problem.

@RequiredArgsConstructor creates constructor only on FINAL fileds. It means String fields will not be affected and it seems better to use already used annotation rather than separate config

Comment (vetterocki) [reply src/main/java/com/qunite/api/security/JwtService.java:18]: `@RequiredArgsConstructor` creates constructor only on FINAL fileds. It means String fields will not be affected and it seems better to use already used annotation rather than separate config

Comment (NikPaukov) [reply src/main/java/com/qunite/api/security/JwtService.java:18]: Final fields is still almost all Fields, i don't like the idea of lazining them without a reason

File: src/main/java/com/qunite/api/web/dto/auth/AuthenticationResponse.java (hunk 2 above)
Comment (NikPaukov) [reply src/main/java/com/qunite/api/web/dto/auth/AuthenticationResponse.java:9]: I do not think that adding noargaconstructor to dto is bad practice

Comment (xd6d) [reply src/main/java/com/qunite/api/web/dto/auth/AuthenticationResponse.java:9]: > I do not think that adding noargaconstructor to dto is bad practice

https://github.com/q-unite/qunite-api/pull/105/files#r1271322615

File: src/main/java/com/qunite/api/data/UserRepository.java
Comment (xd6d) [inline src/main/java/com/qunite/api/data/UserRepository.java:1]: remove unrelated changes

File: src/test/java/com/qunite/api/web/AuthenticationIntegrationTest.java [hunk 9]
```diff
@@ -159,4 +170,59 @@ void signInShouldCheckPassword() throws Exception {
...
+    mockMvc.perform(patch("/users/self").contentType(MediaType.APPLICATION_JSON).content(json)
+            .header(HttpHeaders.AUTHORIZATION, token))
+        .andExpect(status().isOk());
+
+    mockMvc.perform(get("/users/self").header(HttpHeaders.AUTHORIZATION, token))
+        .andExpect(status().isOk());
+  }
+
+  private String getAccessToken(Long id) throws Exception {
+    var user = userService.findOne(id).orElseThrow();
```
Comment (xd6d) [inline src/test/java/com/qunite/api/web/AuthenticationIntegrationTest.java:224]: we have utils method for this

**Review by xd6d — CHANGES_REQUESTED** [review]
I like it, but fix 2 comments
